pkg/reconciler/gc/v2/gc.go

```diff
@@ -61,51 +64,67 @@ func Collect(
 		return a.After(b)
 	})
 
-	for _, rev := range revs[gcSkipOffset:] {
-		if strings.EqualFold(rev.ObjectMeta.Annotations[serving.RevisionPreservedAnnotationKey], "true") {
+	stale, nonactive := 0, 0
+	for _, rev := range revs {
+		if isRevisionActive(rev, config) {
+			continue
+		}
+		nonactive++
+		if isRevisionStale(cfg, rev, logger) {
+			stale++
+		}
+
+		if stale > min {
+			logger.Infof("Deleting stale revision %q", rev.ObjectMeta.Name)
+		} else if max != gc.Disabled && nonactive > max {
+			logger.Infof("Maximum(%d) reached. Deleting oldest non-active revision %q",
+				max, rev.ObjectMeta.Name)
+		} else {
 			continue
 		}
-		if isRevisionStale(ctx, rev, config) {
-			err := client.ServingV1().Revisions(rev.Namespace).Delete(rev.Name, &metav1.DeleteOptions{})
-			if err != nil {
-				logger.With(zap.Error(err)).Errorf("Failed to delete stale revision %q", rev.Name)
-				continue
-			}
+
+		if err := client.ServingV1().Revisions(rev.Namespace).Delete(
```

Comment:
It would be nice if we could use deletecollection 😢

Reply:
https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#set-based-requirement

That would be neat for batch delete (in practice this isn't likely to actually delete more than one since it runs on every deploy), but I think you'd need to do a set selector on `metadata.name`.

 Are set selectors supported for field selectors (rather than labels)? There's no mention of it here: https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors

Reply:
probably we can set a label for things we want to collect and collect them, but probably not worth it.